src/App/Datums.cpp

```diff
@@ -21,42 +23,318 @@
+        return 1;
+    }
+    return DocumentObject::mustExecute();
+}
+
+App::DocumentObjectExecReturn* LocalCoordinateSystem::execute()
+{
+    try {  // try to find all base axis and planes in the origin
+        for (const char* role : AxisRoles) {
+            App::Line* axis = getAxis(role);
+            assert(axis);
+            (void)axis;
+        }
+        for (const char* role : PlaneRoles) {
+            App::Plane* plane = getPlane(role);
+            assert(plane);
+            (void)plane;
+        }
+    }
+    catch (const Base::Exception& ex) {
+        setError();
+        return new App::DocumentObjectExecReturn(ex.what());
+    }
+
+    return DocumentObject::execute();
+}
+
+const std::vector<LocalCoordinateSystem::SetupData>& LocalCoordinateSystem::getSetupData()
+{
+    static const std::vector<SetupData> setupData = {
+        // clang-format off
+        {App::Line::getClassTypeId(),  AxisRoles[0],  tr("X-axis"),   Base::Rotation()},
+        {App::Line::getClassTypeId(),  AxisRoles[1],  tr("Y-axis"),   Base::Rotation(Base::Vector3d(1, 1, 1), M_PI * 2 / 3)},
+        {App::Line::getClassTypeId(),  AxisRoles[2],  tr("Z-axis"),   Base::Rotation(Base::Vector3d(1,-1, 1), M_PI * 2 / 3)},
+        {App::Plane::getClassTypeId(), PlaneRoles[0], tr("XY-plane"), Base::Rotation()},
+        {App::Plane::getClassTypeId(), PlaneRoles[1], tr("XZ-plane"), Base::Rotation(1.0, 0.0, 0.0, 1.0)},
+        {App::Plane::getClassTypeId(), PlaneRoles[2], tr("YZ-plane"), Base::Rotation(Base::Vector3d(1, 1, 1), M_PI * 2 / 3)},
+        {App::Point::getClassTypeId(), PointRoles[0], tr("Origin"),   Base::Rotation()}
+        // clang-format on
+    };
+    return setupData;
+}
+
+DatumElement* LocalCoordinateSystem::createDatum(SetupData& data)
+{
+    App::Document* doc = getDocument();
+    std::string objName = doc->getUniqueObjectName(data.role);
+    App::DocumentObject* featureObj = doc->addObject(data.type.getName(), objName.c_str());
+
+    assert(featureObj && featureObj->isDerivedFrom(App::DatumElement::getClassTypeId()));
+
+    QByteArray byteArray = data.label.toUtf8();
+    featureObj->Label.setValue(byteArray.constData());
+
+    auto* feature = static_cast<App::DatumElement*>(featureObj);
+    feature->Placement.setValue(Base::Placement(Base::Vector3d(), data.rot));
+    feature->Role.setValue(data.role);
+
+    feature->Placement.setStatus(Property::Hidden, true);
+    return feature;
+}
+
+LocalCoordinateSystem::SetupData LocalCoordinateSystem::getData(const char* role)
+{
+    const auto& setupData = getSetupData();
+    for (auto data : setupData) {
+        if (std::strcmp(role, data.role) == 0) {
+            return data;
+        }
+    }
+    return LocalCoordinateSystem::SetupData();
+}
+
+void LocalCoordinateSystem::setupObject()
+{
+    std::vector<App::DocumentObject*> links;
+    const auto& setupData = getSetupData();
+    for (auto data : setupData) {
+        links.push_back(createDatum(data));
+    }
```

Comment:
`std::transform`?

Reply:
I don't understand. You want this :
```
void LocalCoordinateSystem::setupObject()
{
    const auto& setupData = getSetupData();

    std::vector<App::DocumentObject*> links;
    links.reserve(setupData.size()); // Preallocate memory for efficiency

    std::transform(setupData.begin(), setupData.end(), std::back_inserter(links), [this](const auto& data) {
        return createDatum(data);
    });

    OriginFeatures.setValues(links);
}
```
?
If yes why? The syntax is much more complex and less readable.